File: src/test/java/com/gitlab/lae/intellij/actions/tree/ActionNodeTest.java
package com.gitlab.lae.intellij.actions.tree;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.util.Pair;
import org.junit.Test;

import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static javax.swing.KeyStroke.getKeyStroke;
import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;

public final class ActionNodeTest {

    @Test
    public void preparesChildItemForContext() {
        var a = ActionNode.create(
                "a",
                "a",
                null,
                false,
                When.ALWAYS,
                singletonList(getKeyStroke('a')),
                emptyList()
        );
        var b = ActionNode.create(
                "b",
                "b",
                null,
                false,
                When.ALWAYS,
                asList(getKeyStroke('b'), getKeyStroke('z')),
                emptyList()
        );
        var c = ActionNode.create(
                "c",
                "c",
                null,
                false,
                When.ALWAYS,
                singletonList(getKeyStroke('b')),
                emptyList()
        );
        var d = ActionNode.create(
                "d",
                "d",
                null,
                false,
                When.NEVER,
                singletonList(getKeyStroke('b')),
                emptyList()
        );

        var actual = ActionNode.create(
                "id",
                "name",
                null,
                false,
                When.ALWAYS,
                emptyList(),
                asList(a, b, c, d)
        ).prepare(mock(DataContext.class));

        var expected = asList(
                Pair.create(singletonList(getKeyStroke('a')), a),
                Pair.create(singletonList(getKeyStroke('z')), b),
                Pair.create(singletonList(getKeyStroke('b')), c)
        );

        assertEquals(expected, actual);
    }

    @Test
    public void preparesChildItemForContextKeepsEmptyKeyStrokes() {
        var a = ActionNode.create(
                "a",
                "a",
                null,
                false,
                When.ALWAYS,
                emptyList(),
                emptyList()
        );
        var b = ActionNode.create(
                "b",
                "b",
                null,
                false,
                When.ALWAYS,
                singletonList(getKeyStroke('b')),
                emptyList()
        );
        var c = ActionNode.create(
                "c",
                "c",
                null,
                false,
                When.ALWAYS,
                emptyList(),
                emptyList()
        );

        var actual = ActionNode.create(
                "id",
                "name",
                null,
                false,
                When.ALWAYS,
                emptyList(),
                asList(a, b, c)
        ).prepare(mock(DataContext.class));

        var expected = asList(
                Pair.create(emptyList(), a),
                Pair.create(singletonList(getKeyStroke('b')), b),
                Pair.create(emptyList(), c)
        );

        assertEquals(expected, actual);
    }
}

File: src/test/kotlin/com/gitlab/lae/intellij/actions/tree/ActionNodeTest.kt
package com.gitlab.lae.intellij.actions.tree

import com.gitlab.lae.intellij.actions.tree.When.Never
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.kotlin.mock

class ActionNodeTest {

  @Test
  fun `prepares child item for context`() {

    val a = actionNode("a", keys = keys("typed a"))
    val b = actionNode("b", keys = keys("typed b", "typed z"))
    val c = actionNode("c", keys = keys("typed b"))
    val d = actionNode("d", keys = keys("typed b"), condition = Never)

    val actual = actionNode(items = listOf(a, b, c, d)).prepare(mock())
    val expected = listOf(
      keys("typed a") to a,
      keys("typed z") to b,
      keys("typed b") to c,
    )
    assertEquals(expected, actual)
  }

  @Test
  fun `prepares child item for context keeps empty key strokes`() {

    val a = actionNode("a")
    val b = actionNode("b", keys = keys("typed b"))
    val c = actionNode("c")

    val actual = actionNode(items = listOf(a, b, c)).prepare(mock())
    val expected = listOf(
      keys() to a,
      keys("typed b") to b,
      keys() to c,
    )
    assertEquals(expected, actual)
  }
}
